godot-cell/src/guards.rs

```diff
@@ -0,0 +1,185 @@
+    ///
+    /// When the guard is set as non-aliasing, then:
+    ///
+    /// - Any references handed out by this guard must stop accessing the value.
+    ///
+    /// This can be satisfied by using `deref_mut` to get a `&mut T` to the value, then making that reference
+    /// inaccessible for the duration that this guard is marked as non-aliasing. As the rust borrow checker
+    /// will prevent anyone from then using this guard until that reference is dropped.
+    pub(crate) unsafe fn new(
+        state: &'a Mutex<CellState<T>>,
+        count: usize,
+        value: NonNull<T>,
+    ) -> Self {
+        Self {
+            state,
+            count,
+            value,
+        }
+    }
+}
+
+impl<'a, T> Deref for MutGuard<'a, T> {
+    type Target = T;
+
+    fn deref(&self) -> &Self::Target {
+        let count = self.state.lock().unwrap().borrow_state.mut_count();
+        // This is just a best-effort error check. It should never be triggered.
+        assert_eq!(
+            self.count, count,
+            "attempted to access the non-current mutable borrow. **this is a bug, please report it**"
+        );
+        // SAFETY:
+        // It is safe to call `as_ref()` on value when we have a `&self` reference because of the safety
+        // invariants of `new`.
+        unsafe { self.value.as_ref() }
+    }
+}
+
+impl<'a, T> DerefMut for MutGuard<'a, T> {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        let count = self.state.lock().unwrap().borrow_state.mut_count();
+        // This is just a best-effort error check. It should never be triggered.
+        assert_eq!(
+            self.count, count,
+            "attempted to access the non-current mutable borrow. **this is a bug, please report it**"
+        );
+
+        // SAFETY:
+        // It is safe to call `as_mut()` on value when we have a `&mut self` reference because of the safety
+        // invariants of `new`.
+        unsafe { self.value.as_mut() }
+    }
+}
+
+impl<'a, T> Drop for MutGuard<'a, T> {
+    fn drop(&mut self) {
+        self.state
+            .lock()
+            .unwrap()
+            .borrow_state
+            .decrement_mut()
+            .unwrap();
+    }
+}
+
+// ----------------------------------------------------------------------------------------------------------------------------------------------
+
+/// A guard that ensures a mutable reference is kept inaccessible until it is dropped. At which point the
+/// borrow-state is set to prevent aliasing for the most recent borrow and the pointer in state is set to the
+/// previous pointer.
+#[derive(Debug)]
+pub struct NonAliasingGuard<'a, T> {
+    state: &'a Mutex<CellState<T>>,
+    prev_ptr: NonNull<T>,
+}
+
+impl<'a, T> NonAliasingGuard<'a, T> {
+    /// Create a new non-aliasing guard for `state`.
+    pub(crate) fn new(state: &'a Mutex<CellState<T>>, prev_ptr: NonNull<T>) -> Self {
+        Self { state, prev_ptr }
+    }
+}
```

Comment:
It's the fact that we return a value with a lifetime. So since we call this method with a `'a: 'b` i.e `'b` must outlive `'a`, the mutable reference is `&'b mut T`, and we return. This means that the reference must outlive the guard. And since the reference is mutable then we cannot have any references around as well since we captured the lifetime on creation.

I can move some stuff into this function though to make this relationship more obvious.

Reply:
Yes, that would be nice. Please also document the relation/assumption 👍